packages/core/core/src/services/document-service/transform/relations/extract/data-ids.ts

```diff
@@ -47,31 +47,34 @@ const addRelationDocId = curry(
  * Those will later be transformed to entity ids.
  */
 const extractDataIds = (idMap: IdMap, data: Record<string, any>, source: Options) => {
+  // This is not iterating polymorphics
   return traverseEntityRelations(
     async ({ attribute, value }) => {
       if (!attribute) {
         return;
       }
-
-      const targetUid = attribute.target!;
-      const addDocId = addRelationDocId(idMap, targetUid, source);
+      const addDocId = addRelationDocId(idMap, source);
 
       return mapRelation((relation) => {
         if (!relation || !relation.documentId) {
           return relation;
         }
 
-        addDocId(relation);
+        // Regular relations will always target the same target
+        // if its a polymorphic relation we need to get it from the data itself
+        const targetUid = attribute.target || (relation.__type as string);
```

Comment:
Might be secondary, but instead of assuming there will always be a string `__type` property if `attribute.target` is not defined, could we make assertions and raise errors if it's not of the expected type?

Reply:
This way we let TypeScript handles the type narrowing and we keep runtime safety

Reply:
Yeah, I'll clean this up over the next couple days with stuff like that